src/libraries/System.Private.CoreLib/src/System/Buffer.cs

```diff
@@ -182,6 +182,9 @@ internal static unsafe void Memmove<T>(ref T destination, ref T source, nuint el
 #endif
         internal static void BulkMoveWithWriteBarrier(ref byte destination, ref byte source, nuint byteCount)
         {
+            if (byteCount == 0 || Unsafe.AreSame(ref destination, ref source))
```

Comment:
This method seems to be structured to allow inlining of BulkMoveWithWriteBarrier at every callsite.

I think it would be better to keep these checks inside the non-inlined portion of the helpers. The empty buffer or same buffer are rare cases. We do not need to optimize for them by inlining these checks at every callsite.